src/box/lua/config/applier/box_cfg.lua

```diff
@@ -475,7 +444,141 @@ local function apply(config)
+        box.on_commit(function()
+            local watch_key = 'box_apply_names'
+            local names = box.watch_once(watch_key) or {}
+            table.insert(names, new[2])
+            box.broadcast(watch_key, names)
+        end)
+    end
+end
+
+local function names_cluster_on_replace(old, new)
+    if old[3] == nil and new[3] ~= nil then
+        box.on_commit(function()
+            local watch_key = 'box_apply_names'
+            local names = box.watch_once(watch_key) or {}
+            table.insert(names, new[3])
+            box.broadcast(watch_key, names)
+        end)
+    end
+end
+
+local function names_box_cfg(cfg)
+    local ok, err = pcall(function()
+        box.cfg(cfg)
+    end)
+
+    -- We don't consider failed box.cfg a critical error,
+    -- delete a warning anyway. box.info shows correct info.
+    if not ok then
+        log.warn('Failed to apply name in box.cfg: %s', err)
+    end
+end
+
+local function post_apply(config)
+    local configdata = config._configdata
+    if table.equals(configdata._missing_names, {}) then
+        return
+    end
+
+    local names_apply_fiber = fiber.new(names_apply_f, configdata)
+
+    -- Set triggers to notify apply_name watcher
+    box.space._cluster:on_replace(names_cluster_on_replace)
+    box.space._schema:on_replace(names_schema_on_replace)
+
+    local upgrade_on_replace = nil
+    local version = mkversion.get()
+    if version < mkversion(3, 0, 0) then
+        upgrade_on_replace = box.space._schema:on_replace(function(old, new)
+            if old == nil or new == nil then
+                return
+            end
+
+            local version_3 = mkversion(3, 0, 0)
+            local old_version = mkversion.from_tuple(old)
+            local new_version = mkversion.from_tuple(new)
+            if old_version < version_3 and new_version >= version_3 then
+                box.on_commit(function()
+                    pcall(names_apply_fiber.wakeup, names_apply_fiber)
+                end)
+            end
+        end)
+    end
+
+    local watch
+    local config_names = configdata:names()
+    watch = box.watch('box_apply_names', function(_, names)
+        -- The function is executed, when created, even if no broadcasts
+        -- were done. Do nothing in such case.
+        if names == nil then
+            return
+        end
+
+        for _, name in ipairs(names) do
+            if not configdata._missing_names[name] then
+                goto continue
+            end
+            -- Initialize box.cfg.replicaset/instance_name
+            if name == config_names.replicaset_name then
+                names_box_cfg({replicaset_name = name})
+            end
+            if name == config_names.instance_name then
+                names_box_cfg({instance_name = name})
+            end
+
+            -- Delete name from _missing_names and drop the associated alert.
+            configdata._missing_names[name] = nil
+            config:_alert_drop(name)
```

Comment:
It seems removing the name from `_missing_names` should be done from the names_apply_fiber.
Otherwise it may happen that the fiber is woken up before the watcher is, and the fiber sees the name as missing even though it was already replaced

Reply:
The watcher would only do `box.cfg()` then, if needed.

Reply:
I don't really get, what you mean and I don't understand the motivation for such change. For now, I left everything as is. Could you please describe it more extensively?

`names_apply_f` is "woken up" only once, when instance becomes rw. It always wakes up before the watcher. The watcher is actually woken up only once too, after a bunch of replaces take place. So, watcher cannot affect `names_apply_f` in any way

Reply:
In this case maybe I don't understand something.

In my understanding it works as follows:
each time a node receives a replace to _cluster or _schema, it triggers an event with the newly replaced name.
So it may happen that one master replaces one name "name1" (out of many ones missing), then goes ro and another master takes its place. If the fiber is woken up (by `box.ctl.wait_rw()`) before the watcher is signalled that the first master's replace was executed, the fiber will see this particular name "name1" in missing names, even though it's actually already in `_cluster`.

Why's watcher woken up only once? Isn't it signalled every time a replace happens?

P.S. I think I get what you mean now. Since master either replaces all names in a single tx or fails, the watcher will indeed be woken up only once. But I suggest to handle this more carefully in case the user replaces some of the names manually, one by one, or something else happens.
Technically, it's not forbidden to replace the names one by one, so the user, or some other code in the future, may do that. Better be ready for this.

Reply:
Now I see, there's indeed may be a problem. Thank you a lot for explanation. Fixed

Reply:
We don't "persist" `missing_names` anymore